src/Compilers/Core/Portable/FileSystem/PathUtilities.cs

```diff
@@ -443,5 +458,182 @@ internal static bool ContainsPathComponent(string path, string component, bool i
+                {
+                    break;
+                }
+            }
+
+            // if the first part doesn't match, they don't even have the same volume
+            // so there can be no relative path.
+            if (index == 0)
+            {
+                return fullPath;
+            }
+
+            // add backup notation for remaining base path levels beyond the index
+            var remainingParts = directoryPathParts.Length - index;
+            if (remainingParts > 0)
+            {
+                for (int i = 0; i < remainingParts; i++)
+                {
+                    relativePath = relativePath + ParentRelativeDirectory + DirectorySeparatorStr;
+                }
+            }
+
+            // add the rest of the full path parts
+            for (int i = index; i < fullPathParts.Length; i++)
+            {
+                relativePath = CombinePathsUnchecked(relativePath, fullPathParts[i]);
+            }
+
+            return relativePath;
+        }
+
+        /// <summary>
+        /// True if the child path is a child of the parent path.
+        /// </summary>
+        public static bool IsChildPath(string parentPath, string childPath)
+        {
+            return parentPath.Length > 0
+                && childPath.Length > parentPath.Length
+                && PathsEqual(childPath, parentPath, parentPath.Length)
+                && (IsDirectorySeparator(parentPath[parentPath.Length - 1]) || IsDirectorySeparator(childPath[parentPath.Length]));
+        }
+
+        private static string GetRelativeChildPath(string parentPath, string childPath)
+        {
+            var relativePath = childPath.Substring(parentPath.Length);
+
+            // trim any leading separators left over after removing leading directory
+            int start = ConsumeDirectorySeparators(relativePath, relativePath.Length, 0);
+            if (start > 0)
+            {
+                relativePath = relativePath.Substring(start);
+            }
+
+            return relativePath;
+        }
+
+        private static readonly char[] s_pathChars = new char[] { VolumeSeparatorChar, DirectorySeparatorChar, AltDirectorySeparatorChar };
+
+        private static string[] GetPathParts(string path)
+        {
+            var pathParts = path.Split(s_pathChars);
+
+            // remove references to self directories ('.')
+            if (pathParts.Contains(ThisDirectory))
+            {
+                pathParts = pathParts.Where(s => s != ThisDirectory).ToArray();
+            }
+
+            return pathParts;
+        }
+
+        /// <summary>
+        /// True if the two paths are the same.
+        /// </summary>
+        public static bool PathsEqual(string path1, string path2)
+        {
+            return PathsEqual(path1, path2, Math.Max(path1.Length, path2.Length));
+        }
+
+        /// <summary>
+        /// True if the two paths are the same.  (but only up to the specified length)
```

Comment:
Should we document that "equal" in this case means "lexically equal?" Because I know people want that function too and they might be disappointed when they don't get it.